### PR TITLE
Add vicuna dataset

### PR DESCRIPTION
**Pull request by CloseChoice**
#2341

This PR adds the vicuna dataset. We use the filtered vicuna dataset since it only contains english conversations.

The sanity checks are necessary here and filter quite a number of IMO unuseful conversations (where the conversation is empty, where we have less than 2 speakers and where the speakers are not in adjacent order). After all filterings we are left with 84260 of 85071 conversations.

### CODE REVIEW

File: model/model_training/custom_datasets/qa_datasets.py [hunk 1]
```diff
@@ -453,3 +456,64 @@ def __init__(self, mode: str = "sft", cache_dir: str = None) -> None:
...
+        if all(speaker1_idx) and all(speaker2_idx):
+            # [Q1, A1, Q2, A2] -> [Q1, A1, Q2, A2]
+            # Use only input_max_length characters
+            truncated_dialogue = [k["value"][:input_max_length] for k in data["conversations"]]
+            # split dialogue in question and answer pairs.
+            # [Q1, A1, Q2, A2] -> [[Q1, A1], [Q2, A2]]
+            pairs = self._partition_list(truncated_dialogue, 2)
+            # add previous history to conversation pairs
+            # e.g. [[Q1, A1], [Q2, A2]] -> [[Q1, A1], [Q1, A1, Q2, A2]]
+            pairs = list(accumulate(pairs, lambda x, y: x + y))
```
Comment (CloseChoice) [inline model/model_training/custom_datasets/qa_datasets.py:497]: This is not correct. Should be a reduce.

File: model/model_training/custom_datasets/qa_datasets.py [hunk 2]
```diff
@@ -453,3 +456,64 @@ def __init__(self, mode: str = "sft", cache_dir: str = None) -> None:
...
+            if (
+                processed_data := self.process_vicuna_conversations(data, input_max_length=input_max_length)
+            ) is not None:
+                self.pairs.append(processed_data)
+
+    def __len__(self):
+        return len(self.pairs)
+
+    def __getitem__(self, index):
+        return self.pairs[index]
```
Comment (CloseChoice) [inline model/model_training/custom_datasets/qa_datasets.py:519]: TODO: Change return depending on mode

File: model/model_training/custom_datasets/qa_datasets.py (hunk 1 above)
Comment (CloseChoice) [inline model/model_training/custom_datasets/qa_datasets.py:497]: This is not yet correct, should just be the truncated dialogue

Comment (CloseChoice) [reply model/model_training/custom_datasets/qa_datasets.py]: done

File: model/model_training/custom_datasets/qa_datasets.py (hunk 2 above)
Comment (CloseChoice) [reply model/model_training/custom_datasets/qa_datasets.py]: done

File: model/model_training/custom_datasets/__init__.py [hunk 3]
```diff
@@ -45,13 +46,7 @@
...
 ]
 
-RL_DATASETS = [
-    "webgpt",
-    "private_tuning",
-    "alpaca",
-    "hf_summary",
-    "hf_summary_pairs",
-]
+RL_DATASETS = ["webgpt", "private_tuning", "alpaca", "hf_summary", "hf_summary_pairs", "vicuna"]
```
Comment (andreaskoepf) [inline model/model_training/custom_datasets/__init__.py:49]: please use ',' at the end to have one entry per line.

**Review by andreaskoepf — CHANGES_REQUESTED** [review]
looks good, just minor formatting change requested.

**Review by andreaskoepf — APPROVED** [review]

**Pull request merged**